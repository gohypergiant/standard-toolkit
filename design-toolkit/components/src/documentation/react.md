# React

## Best Practices

### Context

RAC utilizes context heavily in their components. Understanding the patterns they've established is critical to understanding how slots and prop inheritance works.

In typical React patterns, context is provided at the top of the application or relatively high in the rendering hierarchy and focuses mainly on global state. However, in RAC, the utilization of context is brought to the component level and often times represents the internal state of the component and / or slots to pass props down to composed components. This means that it's common for components to implement 1 or more contexts.

The other side of this is that you'll often see components consuming these contexts as a first step. This is necessary to maintain correct prop inheritance order. Here's an example:

```jsx
<NumberField>
  <Button slot="decrease">Decrease</Button>
  <Input />
  <Button slot="increase">Increase</Button>
</NumberField>
```

In this scenario, `NumberField` is wrapping it's `children` with multiple contexts, for example purposes we're just going to identify a couple: `ButtonContext` & `InputContext`. From a compositional perspective this is invisible to the implementor except through documentation and reading the source code.

`ButtonContext` is providing two `slot`s, which are named `decrease` & `increase`. Each of these slots are being provide a set of `ButtonProps`, such as aria / data attributes, classNames and event handlers. This makes it so that these buttons "just work" and there's no need for the implementor to bind things up manually. Here's a simplified look at how that looks on both sides:

```jsx
function NumberField({ children, defaultValue, value, onChange }: NumberFieldProps) {
  const [val, setVal] = useControlledState(defaultValue, value, onChange);

  return (
    <ButtonContext.Provider value={{
      slots: {
        decrease: {
          onPress: () => setVal(value - 1)
        }
      }
    }}>
      {children}
    </ButtonContext.Provider>
  )
}

function Button({ ref, ...props }: ButtonProps) {
  [props, ref] = useContextProps(props, ref ?? null, ButtonContext);
}
```

The context providing the slot defines what is necessary to bind to that component in order to have a functional component. The component consuming the context identifies itself with a `slot` prop, pulls in the props off the context and then merges those props with it's own. It's important to note that not all props get merged, some get overridden. The main props being merged are: `ref`, `id`, `className`, `style` and event handlers that follow the pattern `on[A-Z]`, [documented here](https://github.com/adobe/react-spectrum/blob/main/packages/react-aria-components/src/utils.tsx#L183) & [here](https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/utils/src/mergeProps.ts#L36).

If you implement defaults in a component, these definitions must come after `useContextProps`, otherwise it is guaranteed to override and wipe out any props being passed by context (unless they're merged). You should follow this pattern:

```jsx
const MyComponentContext = createContext<ContextValue<MyComponentProps, HTMLDivElement>>(null);

function MyComponent({ ref, ...props }: MyComponentProps) {
  [props, ref] = useContextProps(props, ref ?? null, MyComponentContext);

  const { size = 'medium' } = props;
}
```

Because a prop like `size` isn't one of the merged props it's important that the default get's set after the context consumption. Here's a few different possible scenarios and what the resulting value would be:

- `<MyComponent />` that isn't composed within a context providing one: `medium` (fallback to default implemented in the component)
- `<MyComponent />` that is composed within a context providing `large`: `large` (assuming that the context isn't slotted or that `<MyComponent slot="foo" />`'s slot matches a slot provided in the context)
- `<MyComponent size="small" />`: `small` (doesn't matter if it was composed in a context or not, local overrides always take highest precedence)

### Props

#### Order

For the sake of readability and consistency, prop ordering helps to be able to quickly scan a component to find what you're looking for. The order which follows a sense of uniqueness, priority, and alphabetization is as follows:

- id
- key
- className
- (everything else not present in another group, in alpha order)
- state flags (is[A-Z] in alpha order)
- event handlers (on[A-Z] in alpha order)
- aria attributes (in alpha order)
- data attributes (in alpha order)

What that looks like:

```jsx
<MyComponent
  id="foo"
  key="bar"
  className="foo bar"
  size="small"
  variant="outline"
  isDisabled
  isOpen
  onChange={() => {}}
  onPress={() => {}}
  aria-labeledby="mylabel"
  data-testid="foobar"
/>
```

Having implementations where prop order is scrambled adds undue burden to the reader and often times leads to mistakes and / or bugs. This same ordering is the approach to be taking when destructing props as well, with the only addition being `children` which jumps to the top, and the only caveat being that `ref` needs to be destructured separately for other purposes. Seen here:

```jsx
function MyComponent({ ref, ...props }: MyComponentProps) {
  [props, ref] = useContextProps(props, ref ?? null, MyComponentContext);

  const {
    children,
    id,
    className,
    size,
    variant,
    isDisabled,
    isOpen,
    onChange,
    onPress,
    ...rest,
  } = props;
}
```

#### Spread

When spreading `...rest` on a component, this should be done before the implementation of any props that are being implemented, unless it's acceptable to override them. Here's an example:

```jsx
function MyComponent({ children, ...rest }: MyComponentProps) {
  return (
    <Button {...rest} onPress={() => doSometingImportant()}>
      {children}
    </Button>
  )
}
```

Notice that if the spread of `rest` came afterwards, the `onPress` would be wiped out and the important functionality would never occur. There are rare occurences where the spread will happen in the middle or end of props, but that's the exception to the rule.

### Refs

Make sure to include a ref for the top most element / component being rendered. In your props, include `RefAttributes` and pass the HTML element that matches. If this ref is being passed to an RAC component, the element type will have to match the ref element that RAC has already defined.

```jsx
import type { RefAttributes } from 'react';
import { Button, type ButtonProps } from 'react-aria-components';

type MyComponentProps = ButtonProps & RefAttributes<HTMLButtonElement>;

function MyComponent({ ref }: MyComponentProps) {
  return <Button ref={ref} />
}
```

If developing a more complex component such as a Field where access to an internal element / component may be relevent to the functionality / state of the component, adding additional refs as props is handy. But this should be limited in scope, not every sub element / component should receive a ref.

```jsx
import type { Ref, RefAttributes } from 'react';

type MyComponentProps = RefAttributes<HTMLDivElement> & {
  inputRef?: Ref<HTMLInputElement>
}
```

### Render Props

RAC introduces the concept of render props, where `className`, `style` and/or `children` can be a function provided the internal state of the RAC component as parameters. This is a powerful accessor into that state, but there are some caveats to be aware of:

- Not all components provide the same render props. Certain components will not have `children` as a render prop. This is often times because `children` is serving a differnt function type, typically related to collection rendering.
- When a component is composed within another component, the sub component will not have direct access to the internal state of it's parent, even if it's slotted in. This means that a `FieldError` inside of a `TextField` has different render props and therefore access to a different set of state, even though `TextField` is the state owner in this context.

When developing a component that wraps a component that provides render props, as much as possible, maintain that access to state outwards. Here's what that looks like:

```jsx
import { composeRenderProps, ToggleButton, ToggleButtonProps } from 'react-aria-components';
import { MyComponentStyles } from './styles';

type MyComponentProps = Pick<ToggleButtonProps, 'className' | 'style'>;

function MyComponent({ className }: MyComponentProps) {
  return (
    <ToggleButton
      className={composeRenderProps(className, (className, { isSelected }) =>
        MyComponentStyles({ className, isSelected })
      )}
    />
  );
}
```

This has made it so that your component's styles received the internal state and you've also left the `className` render prop exposed for anybody implementing your component in case they need access to that internal state for their style overrides as well. Keep in mind that it will not always be possible or desirable to maintain all props. The most common scenario where a render prop is replaced is when `children` is removed as a prop due to a specific implementation of sub elements / components.

### Testing

When writing tests for components make sure to follow these dos & donts:

#### DOs

- Create a setup function that cleanly abstracts the initial render
- Test for state through existance of elements, content, or properties
- Prioritize user action based tests over prop configuration tests

#### DONTs

- Test 3rd party code, focus only on code we've developed
- Test for existance of classNames or styles, instead test for state (as described above)
