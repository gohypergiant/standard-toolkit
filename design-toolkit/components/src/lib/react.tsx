/*
 * Copyright 2025 Hypergiant Galactic Systems Inc. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

import {
  Children,
  Fragment,
  type JSXElementConstructor,
  type ReactNode,
  isValidElement,
} from 'react';

class ComponentStructureError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ComponentStructureError';
  }
}

type ChildrenTypes =
  | ReactNode
  | ReactNode[]
  | ((
      // biome-ignore lint/suspicious/noExplicitAny: aria render props include a generic type
      values: any & {
        defaultChildren: ReactNode | undefined;
      },
    ) => ReactNode);

function getChildren(children: ChildrenTypes) {
  const childrenArray = Children.toArray(
    typeof children === 'function'
      ? children({ state: {}, defaultChildren: <></> })
      : children,
  );
  let index = 0;

  while (index < childrenArray.length) {
    const child = childrenArray[index];

    if (
      child != null &&
      typeof child === 'object' &&
      !(child instanceof Promise) &&
      !(Symbol.iterator in child) &&
      child.type === Fragment &&
      child.props != null &&
      typeof child.props === 'object' &&
      'children' in child.props
    ) {
      childrenArray.splice(
        index,
        1,
        ...Children.toArray(child.props.children as ReactNode | ReactNode[]),
      );
    } else {
      index++;
    }
  }

  return childrenArray;
}

type ContainsExactChildrenProps = {
  componentName: string;
  children: ChildrenTypes;
  restrictions: [
    // biome-ignore lint/suspicious/noExplicitAny: allow all props
    string | JSXElementConstructor<any>,
    { min: number; max?: number },
  ][];
};

/**
 * Validates the intended composite component structure.
 *
 * @param children the children of the component
 * @param componentName the displayName of the component
 * @param restrictions the record of validation rules
 */
export function containsExactChildren({
  children,
  componentName,
  restrictions,
}: ContainsExactChildrenProps) {
  if (process.env.NODE_ENV === 'production') {
    return;
  }

  const childrenComponents = getChildren(children);

  if (!childrenComponents.every(isValidElement)) {
    throw new ComponentStructureError(
      `<${componentName}> received invalid children.`,
    );
  }

  const accumulationResults = childrenComponents.reduce(
    (acc: Record<string, number>, child) => {
      const name =
        // @ts-expect-error Accessing undocumented / untyped properties of React components
        child.type?.name ?? child.type?.displayName ?? child.type?.render?.name;

      if (name) {
        acc[name] = (acc[name] || 0) + 1;
      }
      return acc;
    },
    {},
  );

  const missingComponentsArray: string[] = [];
  const excessComponentsArray: string[] = [];

  for (const [component, { min, max }] of restrictions) {
    const name =
      typeof component === 'string'
        ? component
        : // @ts-expect-error Accessing undocumented / untyped properties of React components
          (component.name ?? component.displayName ?? component.render?.name);

    const found = accumulationResults[name] ?? 0;

    if (found < min) {
      missingComponentsArray.push(`${min - found} of <${name}>`);
    }

    if (max !== undefined && found > max) {
      excessComponentsArray.push(`${found - max} of <${name}>`);
    }
  }

  if (missingComponentsArray.length || excessComponentsArray.length) {
    const formatList = (label: string, items: string[]) =>
      items.length ? `\t${label}:\n\t\t${items.join(', ')}\n` : '';

    const errorMessage =
      `Invalid <${componentName}> structure \n` +
      `${formatList('Missing the following', missingComponentsArray)}` +
      `${formatList('Excess of the following', excessComponentsArray)}`;

    throw new ComponentStructureError(errorMessage.trim());
  }
}

/**
 * I look for svgIcons exported from the @accelint/icons library
 * and ensure that they are wrapped in an Icon component in order
 * to get the classes and styles they need in context.
 *
 * Using isValidElement means we will filter out strings, boolean, etc.
 * that are valid nodes but not elements.
 *
 * @example
 *   expectsIconWrapper({
 *     children,
 *     componentName: Button.displayName,
 *   });
 *
 *
 * @param children the children of the calling component
 * @param componentName the displayName of the calling component
 */
export function expectsIconWrapper({
  children,
  componentName,
}: Omit<ContainsExactChildrenProps, 'restrictions'>) {
  if (process.env.NODE_ENV === 'production') {
    return;
  }

  const childrenComponents = Children.toArray(
    children as ReactNode | ReactNode[],
  );

  childrenComponents.map((child) => {
    if (isValidElement(child)) {
      // icons should never be a direct child of the parent
      // @ts-expect-error Accessing undocumented / untyped properties of React components
      if (child.type?.name?.startsWith('Svg')) {
        throw new Error(
          `${componentName} is using an icon without the required Icon wrapper`,
        );
      }
    }
  });
}
