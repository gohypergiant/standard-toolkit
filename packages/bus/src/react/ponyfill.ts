/*
 * Copyright 2025 Hypergiant Galactic Systems Inc. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/**
 * Vendored from https://github.com/sanity-io/use-effect-event/blob/main/src/useEffectEvent.ts
 */

import React, { createContext, useInsertionEffect, useRef } from 'react';

const context = createContext(true);

function forbiddenInRender() {
  throw new Error(
    "A function wrapped in useEffectEvent can't be called during rendering.",
  );
}

// We can only check if we're in a render phase, beyond initial render, in React 19, with its `React.use` hook.
const isInvalidExecutionContextForEventFunction =
  'use' in React
    ? () => {
        // There's no way to check if we're in a render phase from outside of React, the API used by useEffectEvent is private: https://github.com/facebook/react/blob/a00ca6f6b51e46a0ccec54a2231bfe7a1ed9ae1d/packages/react-reconciler/src/ReactFiberWorkLoop.js#L1785-L1788
        // So to emulate the same behavior, we call the use hook and if it doesn't throw, we're in a render phase.
        try {
          return React.use(context);
        } catch {
          return false;
        }
      }
    : () => false;

export function useEffectEvent<const T extends (...args: any[]) => void>(
  fn: T,
): T {
  /**
   * For both React 18 and 19 we set the ref to the forbiddenInRender function, to catch illegal calls to the function during render.
   * Once the insertion effect runs, we set the ref to the actual function.
   */
  const ref = useRef(forbiddenInRender as T);

  useInsertionEffect(() => {
    ref.current = fn;
  }, [fn]);

  return ((...args: any) => {
    // Performs a similar check to what React does for `useEffectEvent`:
    // 1. https://github.com/facebook/react/blob/b7e2de632b2a160bc09edda1fbb9b8f85a6914e8/packages/react-reconciler/src/ReactFiberHooks.js#L2729-L2733
    // 2. https://github.com/facebook/react/blob/b7e2de632b2a160bc09edda1fbb9b8f85a6914e8/packages/react-reconciler/src/ReactFiberHooks.js#L2746C9-L2750
    if (isInvalidExecutionContextForEventFunction()) {
      forbiddenInRender();
    }

    const latestFn = ref.current;
    return latestFn(...args);
  }) as T;
}
