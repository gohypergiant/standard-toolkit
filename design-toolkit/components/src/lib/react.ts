/*
 * Copyright 2025 Hypergiant Galactic Systems Inc. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

// @ts-nocheck
import { Children, type ReactNode, isValidElement } from 'react';

class ComponentStructureError extends Error {
  constructor(message) {
    super(message);
    this.name = 'ComponentStructureError';
  }
}

type ContainsExactChildrenProps = {
  componentName: string;
  // biome-ignore lint/suspicious/noExplicitAny: aria render props include a generic type
  children:
    | ReactNode
    | ReactNode[]
    | ((
        values: any & {
          defaultChildren: ReactNode | undefined;
        },
      ) => ReactNode);
  restrictions: Record<string, { min: number; max?: number }>;
};

/**
 * Validates the intended composite component structure.
 *
 * @param children the children of the component
 * @param componentName the displayName of the component
 * @param restrictions the record of validation rules
 */
export function containsExactChildren({
  children,
  componentName,
  restrictions,
}: ContainsExactChildrenProps) {
  const childrenComponents = Children.toArray(children);

  if (!childrenComponents.every(isValidElement)) {
    throw new ComponentStructureError(
      `<${componentName}> received invalid children.`,
    );
  }

  const accumulationResults = childrenComponents.reduce((acc, child) => {
    const name = child?.type?.displayName;
    if (name) {
      acc[name] = (acc[name] || 0) + 1;
    }
    return acc;
  }, {});

  const missingComponentsArray: string[] = [];
  const excessComponentsArray: string[] = [];

  for (const [key, { min, max }] of Object.entries(restrictions)) {
    const found = accumulationResults[key] ?? 0;

    if (found < min) {
      missingComponentsArray.push(`${min - found} of <${key}>`);
    }

    if (max !== undefined && found > max) {
      excessComponentsArray.push(`${found - max} of <${key}>`);
    }
  }

  if (missingComponentsArray.length || excessComponentsArray.length) {
    const formatList = (label: string, items: string[]) =>
      items.length ? `\t${label}:\n\t\t${items.join(', ')}\n` : '';

    const errorMessage =
      `Invalid <${componentName}> structure \n` +
      `${formatList('Missing the following', missingComponentsArray)}` +
      `${formatList('Excess of the following', excessComponentsArray)}`;

    throw new ComponentStructureError(errorMessage.trim());
  }
}

/**
 * I look for svgIcons exported from the @accelint/icons library
 * and ensure that they are wrapped in an Icon component in order
 * to get the classes and styles they need in context.
 *
 * Using isValidElement means we will filter out strings, boolean, etc.
 * that are valid nodes but not elements.
 *
 * @example
 *   expectsIconWrapper({
 *     children,
 *     componentName: Button.displayName,
 *   });
 *
 *
 * @param children the children of the calling component
 * @param componentName the displayName of the calling component
 */
export function expectsIconWrapper({ children, componentName }) {
  const childrenComponents = Children.toArray(children);

  childrenComponents.map((child) => {
    if (isValidElement(child)) {
      // icons should never be a direct child of the parent
      if (child.type.name?.startsWith('Svg')) {
        throw new Error(
          `${componentName} is using an icon without the required Icon wrapper`,
        );
      }
    }
  });
}
