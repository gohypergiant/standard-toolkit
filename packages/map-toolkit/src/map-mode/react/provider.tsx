/*
 * Copyright 2025 Hypergiant Galactic Systems Inc. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

'use client';

import 'client-only';
import { uuid } from '@accelint/core';
import { createContext, type ReactNode, useEffect, useRef } from 'react';
import { destroyStore, getOrCreateStore } from '../store';
import type { UniqueId } from '@accelint/core';

/**
 * React context for map instance ID.
 * Use the `useMapMode` hook to access the map mode state.
 */
export const MapIdContext = createContext<UniqueId | null>(null);

/**
 * Props for the MapIdProvider component.
 */
export type MapIdProviderProps = {
  /** Child components that will have access to map mode context */
  children: ReactNode;
  /** The default mode to start with. Defaults to 'default' */
  defaultMode?: string;
  /**
   * Optional unique identifier for this map instance.
   *
   * Used to isolate mode changes between different map instances (e.g., main map vs minimap).
   * If not provided, a unique ID is automatically generated.
   *
   * @example
   * ```tsx
   * // Multiple independent map instances
   * <MapIdProvider mapId="main-map-uuid-123">
   *   <BaseMap />
   *   <Toolbar />
   * </MapIdProvider>
   *
   * <MapIdProvider mapId="minimap-uuid-456">
   *   <BaseMap />
   * </MapIdProvider>
   * ```
   */
  mapId?: UniqueId;
};

const DEFAULT_MODE = 'default';

/**
 * Provider component for managing map modes with ownership and authorization.
 *
 * This component uses a hybrid architecture combining React Context (for map instance identity)
 * with an external observable store (for state management). The provider:
 * - Provides a unique `mapInstanceId` via Context
 * - Creates an isolated MapModeStore instance for this map
 * - Allows components to subscribe to mode changes via `useMapMode` hook (which uses `useSyncExternalStore`)
 *
 * The external store manages a state machine for map modes where components can request
 * mode changes with ownership. When a mode is owned by a component, other components
 * must request authorization to change to a different mode. The store handles:
 *
 * - Automatic mode changes when no ownership conflicts exist
 * - Authorization flow when switching between owned modes
 * - Per-mode ownership tracking that persists throughout the session
 * - Pending request management (one pending request per requester)
 * - Auto-acceptance of first pending request when mode owner returns to default
 * - Auto-rejection of other pending requests when one is approved
 * - Event emission through a centralized event bus
 * - Instance isolation for multiple map scenarios (main map + minimap)
 *
 * ## Instance Isolation
 *
 * Each MapIdProvider instance operates independently. Mode changes in one instance
 * do not affect other instances, even when multiple maps are rendered on the same page.
 * This enables scenarios like:
 * - Main map in "drawing" mode while minimap stays in "view" mode
 * - Multiple independent map views with different interaction modes
 *
 * Events are scoped to specific instances using an instanceId that is either provided
 * via the `mapId` prop or auto-generated. The event bus filters events to ensure each
 * provider only responds to events for its own instance.
 *
 * ## Pending Request Behavior
 *
 * - Pending requests are stored by requester ID (not mode owner)
 * - Each requester can have only one pending request at a time
 * - New requests from the same requester auto-replace previous requests
 * - Pending requests persist when mode owner switches between their own modes
 * - When any request is approved, all other pending requests are auto-rejected
 * - When mode owner returns to default mode:
 *   - If first pending request is for default mode, all pending requests are rejected (already in requested mode)
 *   - If first pending request is for a different mode, that request is auto-approved and others are rejected
 *
 * @param props - Provider props including children, optional defaultMode, and optional mapId
 * @returns Provider component that wraps children with map instance identity context
 *
 * @example
 * Basic usage:
 * ```tsx
 * import { MapIdProvider } from '@accelint/map-toolkit/map-mode/react';
 *
 * function App() {
 *   return (
 *     <MapIdProvider defaultMode="view">
 *       <MapView />
 *     </MapIdProvider>
 *   );
 * }
 * ```
 *
 * @example
 * ```tsx
 * // Multiple independent map instances with unique IDs
 * <MapIdProvider mapId="main-map-uuid-abc" defaultMode="drawing">
 *   <BaseMap />
 *   <Toolbar />
 * </MapIdProvider>
 *
 * <MapIdProvider mapId="minimap-uuid-xyz" defaultMode="view">
 *   <BaseMap />
 * </MapIdProvider>
 *
 * // With authorization handling - include mapInstanceId
 * useOn(MapModeEvents.changeAuthorization, (event) => {
 *   const { authId, mapInstanceId } = event.payload;
 *   emitDecision({ authId, approved: true, owner: 'tool', mapInstanceId });
 * });
 * ```
 */
export function MapIdProvider({
  children,
  defaultMode = DEFAULT_MODE,
  mapId,
}: MapIdProviderProps) {
  // Generate stable instance ID (only once per provider mount)
  const autoIdRef = useRef<UniqueId>(null);
  if (!autoIdRef.current) {
    autoIdRef.current = uuid();
  }
  const mapInstanceId: UniqueId = mapId ?? autoIdRef.current;

  // Create or get the store for this map instance synchronously
  // This must happen before any child components try to use useMapMode
  const storeRef = useRef<boolean>(false);
  if (!storeRef.current) {
    getOrCreateStore(mapInstanceId, defaultMode);
    storeRef.current = true;
  }

  // Cleanup: destroy store when provider unmounts
  useEffect(() => {
    return () => {
      destroyStore(mapInstanceId);
    };
  }, [mapInstanceId]);

  return (
    <MapIdContext.Provider value={mapInstanceId}>
      {children}
    </MapIdContext.Provider>
  );
}
