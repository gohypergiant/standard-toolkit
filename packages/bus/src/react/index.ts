// __private-exports
/*
 * Copyright 2025 Hypergiant Galactic Systems Inc. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

import { useEffect, useRef } from 'react';
import { Broadcast } from '../broadcast';
import { useEffectEvent } from './ponyfill';
import type { ExtractEvent, Payload } from '../broadcast/types';

/**
 * A convenience wrapper for useEmit & useOn, to pass down types instead of having
 * to reimplement generics each time
 */
export function useBus<P extends Payload = Payload>() {
  return {
    useEmit: useEmit<P>,
    useOn: useOn<P>,
  };
}

/**
 * React hook to enable render-safe emitting of event with payload that is type safe
 * @template P union of event types
 * @template T type of event
 * @param type of type T, one of the event types
 * @returns callback that will accept the cooresponding payload to the previously entered event type
 */
export function useEmit<
  P extends Payload = Payload,
  T extends P['type'] = P['type'],
>(type: T) {
  const bus = useRef(Broadcast.getInstance<P>());

  return useEffectEvent((payload: ExtractEvent<P, T>['payload']) => {
    bus.current.emit(type, payload);
  });
}

/**
 * React hook to attach event bus listener with type safe callback
 * @param type event type
 * @param callback handler that matches event type and receives cooresponding payload
 */
export function useOn<
  P extends Payload = Payload,
  T extends P['type'] = P['type'],
>(type: T, callback: (data: ExtractEvent<P, T>) => void) {
  const bus = useRef(Broadcast.getInstance<P>());
  const onCallback = useEffectEvent(callback);

  // biome-ignore lint/correctness/useExhaustiveDependencies: onCallback is stable
  useEffect(() => bus.current.on(type, onCallback), [type]);
}
