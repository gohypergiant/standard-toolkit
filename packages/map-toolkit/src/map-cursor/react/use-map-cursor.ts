import { useContext, useMemo, useSyncExternalStore } from 'react';
import { getStore } from '../store';
import { MapIdContext } from './provider';
import type { UniqueId } from '@accelint/core';

/**
 * Return value for the useMapMode hook
 */
export type UseMapCursorReturn = {
  /** The current active map mode */
  cursors: [];
  /** Function to request a mode change with ownership */
  requestCursorChange: (desiredMode: string, requestOwner: string) => void;
};

export function useMapMode(mapInstanceId?: UniqueId): UseMapCursorReturn {
  const contextId = useContext(MapIdContext);
  const actualId = mapInstanceId ?? contextId;

  if (!actualId) {
    throw new Error(
      'useMapCursor requires either a mapInstanceId parameter or to be used within a MapIdProvider',
    );
  }

  // Get the store for this map instance
  const store = getStore(actualId);

  if (!store) {
    throw new Error(
      `MapCursorStore not found for instance: ${actualId}. Ensure a store has been created for this map instance (e.g., via MapIdProvider or getOrCreateStore).`,
    );
  }

  // Subscribe to store using useSyncExternalStore
  const cursors = useSyncExternalStore(store.subscribe, store.getSnapshot);

  // Memoize the return value to prevent unnecessary re-renders
  return useMemo(
    () => ({
      cursors,
      requestCursorChange: store.requestCursorChange,
    }),
    [cursors, store],
  );
}
