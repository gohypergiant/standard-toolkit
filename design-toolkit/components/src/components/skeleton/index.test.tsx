/*
 * Copyright 2025 Hypergiant Galactic Systems Inc. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { Skeleton } from '.';

const baseClasses = (element: HTMLElement) => {
  expect(element).toHaveClass('fg-default-light');
  expect(element).toHaveClass('bg-transparent-light');
  expect(element).toHaveClass('px-l');
  expect(element).toHaveClass('py-s');
  expect(element).toHaveClass('motion-safe:animate-pulse');
};

const circClasses = (element: HTMLElement) => {
  expect(element).toHaveClass('aspect-square');
  expect(element).toHaveClass('h-[1lh]');
  expect(element).toHaveClass('rounded-full');
};

const rectClasses = (element: HTMLElement) => {
  expect(element).toHaveClass('min-h-[1lh]');
  expect(element).toHaveClass('w-full');
  expect(element).toHaveClass('rounded-small');
};

describe('Skeleton', () => {
  it('should render with default classes', () => {
    const { container } = render(<Skeleton />);
    const element = container.firstChild as HTMLElement;

    baseClasses(element);
    rectClasses(element);
  });

  it('should render with children', () => {
    const { container } = render(<Skeleton>Loading content...</Skeleton>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with circle shape', () => {
    const { container } = render(<Skeleton shape='circ' />);
    const element = container.firstChild as HTMLElement;

    baseClasses(element);
    circClasses(element);
  });

  it('should render with more than one and nested', () => {
    const { container } = render(
      <div>
        <Skeleton className='mb-xl p-m'>
          <br />
          <br />
          <Skeleton className='w-[75%]' />
        </Skeleton>

        <Skeleton className='mt-l w-[50%]' />
        <Skeleton className='mt-l w-[50%]' />
        <Skeleton className='mt-l w-[50%]' />
        <Skeleton className='mt-l w-[50%]' />
      </div>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with custom className', () => {
    const { container } = render(<Skeleton className='custom-class' />);
    const element = container.firstChild as HTMLElement;

    expect(element).toHaveClass('custom-class');
    baseClasses(element);
    rectClasses(element);
  });
});
