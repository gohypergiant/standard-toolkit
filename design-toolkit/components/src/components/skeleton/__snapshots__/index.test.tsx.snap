// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Skeleton > should render with children 1`] = `
<div
  class="fg-default-light bg-transparent-light px-l py-s motion-safe:animate-pulse min-h-[1lh] w-full rounded-small"
>
  Loading content...
</div>
`;

exports[`Skeleton > should render with more than one and nested 1`] = `
<div>
  <div
    class="fg-default-light bg-transparent-light motion-safe:animate-pulse min-h-[1lh] w-full rounded-small mb-xl p-m"
  >
    <br />
    <br />
    <div
      class="fg-default-light bg-transparent-light px-l py-s motion-safe:animate-pulse min-h-[1lh] rounded-small w-[75%]"
    />
  </div>
  <div
    class="fg-default-light bg-transparent-light px-l py-s motion-safe:animate-pulse min-h-[1lh] rounded-small mt-l w-[50%]"
  />
  <div
    class="fg-default-light bg-transparent-light px-l py-s motion-safe:animate-pulse min-h-[1lh] rounded-small mt-l w-[50%]"
  />
  <div
    class="fg-default-light bg-transparent-light px-l py-s motion-safe:animate-pulse min-h-[1lh] rounded-small mt-l w-[50%]"
  />
  <div
    class="fg-default-light bg-transparent-light px-l py-s motion-safe:animate-pulse min-h-[1lh] rounded-small mt-l w-[50%]"
  />
</div>
`;
